src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.cs

```diff
@@ -0,0 +1,244 @@
+            Workspace workspace,
+            DiagnosticReporter diagnosticReporter,
+            ProjectFileLoaderRegistry projectFileLoaderRegistry,
+            ImmutableDictionary<string, string> properties)
+        {
+            _workspace = workspace;
+            _diagnosticReporter = diagnosticReporter ?? new DiagnosticReporter(workspace);
+            _pathResolver = new PathResolver(_diagnosticReporter);
+            _projectFileLoaderRegistry = projectFileLoaderRegistry ?? new ProjectFileLoaderRegistry(workspace, _diagnosticReporter);
+            _properties = properties ?? ImmutableDictionary<string, string>.Empty;
+
+            BuildManager = new ProjectBuildManager();
+        }
+
+        /// <summary>
+        /// Create a new instance of an <see cref="MSBuildProjectLoader"/>.
+        /// </summary>
+        public MSBuildProjectLoader(Workspace workspace, ImmutableDictionary<string, string> properties = null)
+            : this(workspace, diagnosticReporter: null, projectFileLoaderRegistry: null, properties)
+        {
+        }
+
+        /// <summary>
+        /// The MSBuild properties used when interpreting project files.
+        /// These are the same properties that are passed to msbuild via the /property:&lt;n&gt;=&lt;v&gt; command line argument.
+        /// </summary>
+        public ImmutableDictionary<string, string> Properties => _properties;
+
+        /// <summary>
+        /// Determines if metadata from existing output assemblies is loaded instead of opening referenced projects.
+        /// If the referenced project is already opened, the metadata will not be loaded.
+        /// If the metadata assembly cannot be found the referenced project will be opened instead.
+        /// </summary>
+        public bool LoadMetadataForReferencedProjects { get; set; } = false;
+
+        /// <summary>
+        /// Determines if unrecognized projects are skipped when solutions or projects are opened.
+        /// 
+        /// A project is unrecognized if it either has 
+        ///   a) an invalid file path, 
+        ///   b) a non-existent project file,
+        ///   c) has an unrecognized file extension or 
+        ///   d) a file extension associated with an unsupported language.
+        /// 
+        /// If unrecognized projects cannot be skipped a corresponding exception is thrown.
+        /// </summary>
+        public bool SkipUnrecognizedProjects { get; set; } = true;
+
+        /// <summary>
+        /// Associates a project file extension with a language name.
+        /// </summary>
+        public void AssociateFileExtensionWithLanguage(string projectFileExtension, string language)
+        {
+            if (language == null)
+            {
+                throw new ArgumentNullException(nameof(language));
+            }
+
+            if (projectFileExtension == null)
+            {
+                throw new ArgumentNullException(nameof(projectFileExtension));
+            }
```

Comment:
>} [](start = 12, length = 1)

Nit: shouldn't argument validation be ordered same as parameters? I'll go back to my OCD corner now.

Reply:
No worries! This particular argument validation ordering already existed in MSBuildWorkspace and I left it as is. I'll flip it.

Nerd.